### PR TITLE
Test bulk upsert stuff

### PR DESCRIPTION
**Pull request by AlexeyRaga**
## Changes

- Refactored BulkUpsert API
- Support plain and scripted bulk upserts
- Wrote integration tests for bulk upsert (scripted and not)

### CODE REVIEW

File: src/Database/Bloodhound/Client.hs [hunk 1]
```diff
@@ -995,12 +995,21 @@ encodeBulkOperation (BulkUpdate (IndexName indexName)
 encodeBulkOperation (BulkUpsert (IndexName indexName)
```
Comment (joshgodsiff) [inline src/Database/Bloodhound/Client.hs:995]: Are we using this function externally? The changes have altered the API.

**Comment by joshgodsiff** [comment]
👍

File: src/Database/Bloodhound/Client.hs (hunk 1 above)
Comment (AlexeyRaga) [reply src/Database/Bloodhound/Client.hs:995]: We do _construct_ `BulkUpsert` externally, yes. 
It is an application code, and that code will have to change.

Comment (joshgodsiff) [reply src/Database/Bloodhound/Client.hs:995]: Fair enough. I doubt we were using the changed bits, anyway, and we're going to touch all those projects.

**Pull request merged**